File: 코드/ch6/clip4/boj2661/main.py
import sys

def solve(end_index):
    if end_index == n:
        for i in range(n):
            print(numbers[i], end='')
        return True
    else:
        for i in range(1, 4):
            numbers[end_index] = i
            if not is_bad(end_index):
                if solve(end_index + 1):
                    return True
    return False

def is_bad(end_index):
    for i in range(1, (end_index + 1) // 2 + 1):
        is_same = True
        for j in range(i):
            if numbers[end_index - j] != numbers[end_index - i - j]:
                is_same = False
                break
        if is_same:
            return True
    return False

n = int(sys.stdin.readline())
numbers = [0] * 80
solve(0)

File: 코드/ch6/clip4/boj2661/Main.java
import java.util.Scanner;

public class Main {
    public static int n;

    public static boolean solve(int endIndex) {
        if (endIndex == n) {
            for (int i = 0; i < n; i++) {
                System.out.print(numbers[i]);
            }
            return true;
        } else {
            for (int i = 1; i <= 3; i++) {
                numbers[endIndex] = i;
                if (!isBad(endIndex)) {
                    if (solve(endIndex + 1)) return true;
                }
            }
        }
        return false;
    }

    public static int[] numbers = new int[80];

    public static boolean isBad(int endIndex) {
        for (int i = 1; i <= (endIndex + 1) / 2; i++) {
            boolean isSame = true;
            for (int j = 0; j < i; j++) {
                if (numbers[endIndex - j] != numbers[endIndex - i - j]) {
                    isSame = false;
                    break;
                }
            }
            if (isSame) return true;
        }
        return false;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        solve(0);
    }
}
